lib/jwt/encode.rb

```diff
@@ -39,7 +39,7 @@ def encoded_header_and_payload
     end
 
     def encode_header
-      encode(@headers.merge(ALG_KEY => @algorithm))
+      encode(@headers.transform_keys(&:to_s).merge(ALG_KEY => @algorithm))
```

Comment:
I think Hash#transform_keys was added to Ruby 2.5.0, so this won't work on older rubies. I think something like what is done [here](https://github.com/jwt/ruby-jwt/blob/master/lib/jwt/claims_validator.rb#L12) could be applied for the headers also.

Also if a copy of the ```@headers``` hash is created the ```Hash#merge!``` method could be used to save one object allocation.

Reply:
I've moved the stringify'ing (as done in the claims validator) to the constructor, and directly add `ALG_KEY` to the headers hash instead of `.merge`

Reply:
Nice, looks good to me.